File: brain/brain.py
#!/usr/bin/env python3
"""
brain.py — Мозг MangoGeneration (Python 3.14+)
Генерирует изображения, анализирует файлы для сортировки, конвертирует форматы.
Работает как локальный скрипт, вызываемый из Rust через процесс или pyo3.
"""

import sys
import os
import json
import math
import time
from pathlib import Path
from typing import Optional

# Пытаемся импортировать Pillow (нужен для генерации изображений)
try:
    from PIL import Image, ImageDraw, ImageFilter
    HAS_PILLOW = True
except ImportError:
    HAS_PILLOW = False
    print("brain: Pillow не установлен. Генерация изображений недоступна.", file=sys.stderr)
    print("brain: Установите: pip install Pillow", file=sys.stderr)


# === МОДУЛЬ ГЕНЕРАЦИИ ИЗОБРАЖЕНИЙ ===

def generate_gradient(
    width: int = 1920,
    height: int = 1080,
    color_start: tuple = (30, 15, 60),
    color_end: tuple = (255, 100, 50),
    direction: str = "vertical",
    output_path: str = "wallpaper.png"
) -> str:
    """
    Генерирует градиентное изображение.
    
    Args:
        width: Ширина изображения
        height: Высота изображения
        color_start: Начальный цвет (R, G, B)
        color_end: Конечный цвет (R, G, B)
        direction: Направление градиента ("vertical", "horizontal", "diagonal")
        output_path: Путь для сохранения
    
    Returns:
        Путь к сохранённому файлу
    """
    if not HAS_PILLOW:
        raise RuntimeError("Pillow не установлен")
    
    img = Image.new("RGB", (width, height))
    pixels = img.load()
    
    for y in range(height):
        for x in range(width):
            # Вычисляем прогресс для разных направлений
            if direction == "vertical":
                t = y / height
            elif direction == "horizontal":
                t = x / width
            elif direction == "diagonal":
                t = (x / width + y / height) / 2
            else:
                t = y / height
            
            # Линейная интерполяция цветов
            r = int(color_start[0] + (color_end[0] - color_start[0]) * t)
            g = int(color_start[1] + (color_end[1] - color_start[1]) * t)
            b = int(color_start[2] + (color_end[2] - color_start[2]) * t)
            
            pixels[x, y] = (r, g, b)
    
    img.save(output_path)
    return output_path


def generate_abstract_pattern(
    width: int = 1920,
    height: int = 1080,
    seed: Optional[int] = None,
    output_path: str = "pattern.png"
) -> str:
    """
    Генерирует абстрактный паттерн из полупрозрачных кругов.
    
    Args:
        width: Ширина
        height: Высота
        seed: Зерно генератора (для воспроизводимости)
        output_path: Путь сохранения
    
    Returns:
        Путь к файлу
    """
    if not HAS_PILLOW:
        raise RuntimeError("Pillow не установлен")
    
    import random
    if seed is not None:
        random.seed(seed)
    else:
        random.seed(int(time.time()))
    
    img = Image.new("RGB", (width, height), (10, 10, 20))
    overlay = Image.new("RGBA", (width, height), (0, 0, 0, 0))
    draw = ImageDraw.Draw(overlay)
    
    # Рисуем случайные полупрозрачные круги
    num_circles = random.randint(20, 60)
    for _ in range(num_circles):
        x = random.randint(-100, width + 100)
        y = random.randint(-100, height + 100)
        radius = random.randint(50, 300)
        r = random.randint(0, 255)
        g = random.randint(0, 255)
        b = random.randint(0, 255)
        alpha = random.randint(30, 100)
        
        draw.ellipse(
            [x - radius, y - radius, x + radius, y + radius],
            fill=(r, g, b, alpha)
        )
    
    # Накладываем и размываем
    img = img.convert("RGBA")
    img = Image.alpha_composite(img, overlay)
    img = img.filter(ImageFilter.GaussianBlur(radius=15))
    img = img.convert("RGB")
    
    img.save(output_path)
    return output_path


def generate_dark_wallpaper(
    width: int = 1920,
    height: int = 1080,
    output_path: str = "dark_wallpaper.png"
) -> str:
    """Генерирует тёмные обои для вечернего режима."""
    return generate_gradient(
        width=width,
        height=height,
        color_start=(10, 5, 30),
        color_end=(40, 20, 80),
        direction="diagonal",
        output_path=output_path
    )


# === МОДУЛЬ АНАЛИЗА И СОРТИРОВКИ ФАЙЛОВ ===

# Правила сортировки по расширению
SORT_RULES = {
    # Изображения
    "jpg": "Images", "jpeg": "Images", "png": "Images",
    "gif": "Images", "bmp": "Images", "webp": "Images",
    "svg": "Images", "tiff": "Images",
    
    # Документы
    "pdf": "Documents/PDF", "doc": "Documents/Word",
    "docx": "Documents/Word", "txt": "Documents/Text",
    "odt": "Documents/Text", "rtf": "Documents/Text",
    
    # Таблицы
    "xls": "Documents/Spreadsheets", "xlsx": "Documents/Spreadsheets",
    "csv": "Documents/Spreadsheets",
    
    # Видео
    "mp4": "Video", "avi": "Video", "mkv": "Video",
    "mov": "Video", "wmv": "Video", "webm": "Video",
    
    # Аудио
    "mp3": "Audio", "wav": "Audio", "flac": "Audio",
    "ogg": "Audio", "aac": "Audio", "m4a": "Audio",
    
    # Архивы
    "zip": "Archives", "tar": "Archives", "gz": "Archives",
    "rar": "Archives", "7z": "Archives",
    
    # Код
    "py": "Code/Python", "js": "Code/JavaScript",
    "ts": "Code/TypeScript", "rs": "Code/Rust",
    "go": "Code/Go", "lua": "Code/Lua",
    "c": "Code/C", "cpp": "Code/C++", "h": "Code/Headers",
}


def analyze_file(filepath: str) -> dict:
    """
    Анализирует файл и предлагает категорию для сортировки.
    
    Args:
        filepath: Путь к файлу
    
    Returns:
        Словарь с информацией о файле
    """
    path = Path(filepath)
    ext = path.suffix.lstrip(".").lower()
    size = path.stat().st_size if path.exists() else 0
    
    # Определяем категорию
    category = SORT_RULES.get(ext, "Other")
    
    # Определяем приоритет (чем меньше расширение — тем выше)
    priority_map = {
        "pdf": 1, "doc": 1, "docx": 1,
        "jpg": 2, "jpeg": 2, "png": 2,
        "mp4": 3, "avi": 3, "mkv": 3,
        "mp3": 4, "wav": 4, "flac": 4,
    }
    
    return {
        "path": str(path.absolute()),
        "name": path.name,
        "extension": ext,
        "size_bytes": size,
        "size_human": _format_size(size),
        "category": category,
        "priority": priority_map.get(ext, 99),
        "suggested_folder": category.split("/")[-1],
    }


def _format_size(size_bytes: int) -> str:
    """Форматирует размер в читаемый вид."""
    for unit in ["B", "KB", "MB", "GB", "TB"]:
        if size_bytes < 1024:
            return f"{size_bytes:.1f} {unit}"
        size_bytes /= 1024
    return f"{size_bytes:.1f} PB"


def analyze_directory(dirpath: str, recursive: bool = True) -> list:
    """
    Анализирует все файлы в директории.
    
    Args:
        dirpath: Путь к директории
        recursive: Рекурсивно ли обходить поддиректории
    
    Returns:
        Список результатов анализа
    """
    results = []
    path = Path(dirpath)
    
    if not path.exists():
        return results
    
    pattern = "**/*" if recursive else "*"
    
    for item in path.glob(pattern):
        if item.is_file():
            try:
                info = analyze_file(str(item))
                results.append(info)
            except (PermissionError, OSError) as e:
                print(f"brain: Ошибка анализа {item}: {e}", file=sys.stderr)
    
    return results


# === МОДУЛЬ КОНВЕРТАЦИИ ФОРМАТОВ ===

def convert_image(
    input_path: str,
    output_path: str,
    output_format: str = "PNG",
    quality: int = 95
) -> str:
    """
    Конвертирует изображение в другой формат.
    
    Args:
        input_path: Путь к исходному файлу
        output_path: Путь к результату
        output_format: Целевой формат (PNG, JPEG, WEBP, BMP)
        quality: Качество (для JPEG/WEBP)
    
    Returns:
        Путь к конвертированному файлу
    """
    if not HAS_PILLOW:
        raise RuntimeError("Pillow не установлен")
    
    img = Image.open(input_path)
    
    # Конвертируем в RGB если нужно сохранить в JPEG
    if output_format.upper() in ("JPEG", "JPG") and img.mode in ("RGBA", "P"):
        img = img.convert("RGB")
    
    # Сохраняем
    save_kwargs = {}
    if output_format.upper() in ("JPEG", "JPG", "WEBP"):
        save_kwargs["quality"] = quality
    
    img.save(output_path, format=output_format.upper(), **save_kwargs)
    return output_path


# === ТОЧКА ВХОДА ДЛЯ CLI ===

def main():
    """CLI-интерфейс brain.py для вызова из Rust."""
    if len(sys.argv) < 2:
        print("Использование:")
        print("  python brain.py generate [--type gradient|pattern|dark] [--output file.png]")
        print("  python brain.py analyze <путь_к_файлу_или_папке>")
        print("  python brain.py convert <input> <output> [--format PNG|JPEG|WEBP]")
        sys.exit(1)
    
    command = sys.argv[1]
    
    if command == "generate":
        wallpaper_type = "gradient"
        output = "wallpaper.png"
        
        i = 2
        while i < len(sys.argv):
            if sys.argv[i] == "--type" and i + 1 < len(sys.argv):
                wallpaper_type = sys.argv[i + 1]
                i += 2
            elif sys.argv[i] == "--output" and i + 1 < len(sys.argv):
                output = sys.argv[i + 1]
                i += 2
            else:
                i += 1
        
        if wallpaper_type == "gradient":
            result = generate_gradient(output_path=output)
        elif wallpaper_type == "pattern":
            result = generate_abstract_pattern(output_path=output)
        elif wallpaper_type == "dark":
            result = generate_dark_wallpaper(output_path=output)
        else:
            print(f"Неизвестный тип: {wallpaper_type}", file=sys.stderr)
            sys.exit(1)
        
        print(json.dumps({"status": "ok", "path": result}))
    
    elif command == "analyze":
        if len(sys.argv) < 3:
            print("Укажите путь для анализа", file=sys.stderr)
            sys.exit(1)
        
        target = sys.argv[2]
        if os.path.isdir(target):
            results = analyze_directory(target)
        else:
            results = [analyze_file(target)]
        
        print(json.dumps(results, ensure_ascii=False, indent=2))
    
    elif command == "convert":
        if len(sys.argv) < 4:
            print("Укажите input и output", file=sys.stderr)
            sys.exit(1)
        
        input_path = sys.argv[2]
        output_path = sys.argv[3]
        fmt = "PNG"
        
        if "--format" in sys.argv:
            idx = sys.argv.index("--format")
            if idx + 1 < len(sys.argv):
                fmt = sys.argv[idx + 1]
        
        result = convert_image(input_path, output_path, fmt)
        print(json.dumps({"status": "ok", "path": result}))
    
    else:
        print(f"Неизвестная команда: {command}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: brain/brain.c
/*
 * brain.c — Мозг MangoGeneration (C11)
 *
 * Локальный генератор и преобразователь изображений + анализатор файлов.
 * Работает как автономный исполняемый файл, вызываемый из Rust-ядра:
 *
 *   brain generate --type gradient|pattern|dark [--width W] [--height H] [--output FILE.png]
 *   brain analyze <путь_к_файлу_или_папке>
 *   brain convert <input> <output> [--format PNG|JPEG|BMP|TGA]
 *
 * Выводит JSON на stdout. Собран с публичными заголовками stb (public domain)
 * для декодирования/кодирования изображений.
 */

/* Должно быть определено ДО любых системных заголовков (для realpath/strcasecmp) */
#ifndef _WIN32
#define _POSIX_C_SOURCE 200809L
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <stdarg.h>
#include <time.h>

/* Кроссплатформенная работа с файловой системой */
#ifdef _WIN32
#include <windows.h>
#include <direct.h>
#define strcasecmp _stricmp
#else
#include <strings.h>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#endif

/* Один исходный файл объявляет реализацию stb */
#define STB_IMAGE_IMPLEMENTATION
#define STB_IMAGE_WRITE_IMPLEMENTATION
#include "vendor/stb_image.h"
#include "vendor/stb_image_write.h"

/* ============================================================
 *  Динамическая строка (простой буфер для сборки JSON)
 * ============================================================ */

typedef struct {
    char *data;
    size_t len;
    size_t cap;
} StrBuf;

static void sb_init(StrBuf *sb) {
    sb->cap = 256;
    sb->len = 0;
    sb->data = (char *)malloc(sb->cap);
    if (sb->data) sb->data[0] = '\0';
}

static void sb_reserve(StrBuf *sb, size_t extra) {
    if (sb->len + extra + 1 > sb->cap) {
        size_t ncap = sb->cap ? sb->cap : 256;
        while (ncap < sb->len + extra + 1) ncap *= 2;
        sb->data = (char *)realloc(sb->data, ncap);
        sb->cap = ncap;
    }
}

static void sb_append(StrBuf *sb, const char *s) {
    size_t l = strlen(s);
    sb_reserve(sb, l);
    memcpy(sb->data + sb->len, s, l);
    sb->len += l;
    sb->data[sb->len] = '\0';
}

static void sb_append_char(StrBuf *sb, char c) {
    sb_reserve(sb, 1);
    sb->data[sb->len++] = c;
    sb->data[sb->len] = '\0';
}

static void sb_printf(StrBuf *sb, const char *fmt, ...) {
    char tmp[4096];
    va_list ap;
    va_start(ap, fmt);
    int n = vsnprintf(tmp, sizeof(tmp), fmt, ap);
    va_end(ap);
    if (n <= 0) return;
    sb_reserve(sb, (size_t)n);
    memcpy(sb->data + sb->len, tmp, (size_t)n);
    sb->len += (size_t)n;
    sb->data[sb->len] = '\0';
}

/* JSON-экранирование строки (без UTF-8 транскодирования) */
static void sb_append_json_string(StrBuf *sb, const char *s) {
    sb_append_char(sb, '"');
    for (const unsigned char *p = (const unsigned char *)s; *p; p++) {
        switch (*p) {
            case '"':  sb_append(sb, "\\\""); break;
            case '\\': sb_append(sb, "\\\\"); break;
            case '\n': sb_append(sb, "\\n");  break;
            case '\r': sb_append(sb, "\\r");  break;
            case '\t': sb_append(sb, "\\t");  break;
            default:
                if (*p < 0x20) sb_printf(sb, "\\u%04x", *p);
                else sb_append_char(sb, (char)*p);
        }
    }
    sb_append_char(sb, '"');
}

/* ============================================================
 *  Случайный генератор (xorshift64 — быстрый и детерминированный)
 * ============================================================ */

static uint64_t rng_state = 0x9E3779B97F4A7C15ULL;

static uint32_t brain_rand(void) {
    rng_state ^= rng_state >> 12;
    rng_state ^= rng_state << 25;
    rng_state ^= rng_state >> 27;
    return (uint32_t)((rng_state * 0x2545F4914F6CDD1DULL) >> 32);
}

static int brain_rand_range(int lo, int hi) {
    if (hi <= lo) return lo;
    return lo + (int)(brain_rand() % (uint32_t)(hi - lo + 1));
}

/* ============================================================
 *  Генерация изображений (RGB-буфер -> PNG через stb)
 * ============================================================ */

static unsigned char *make_gradient(int w, int h, int c1[3], int c2[3],
                                    const char *direction) {
    unsigned char *img = (unsigned char *)malloc((size_t)w * h * 3);
    if (!img) return NULL;

    for (int y = 0; y < h; y++) {
        for (int x = 0; x < w; x++) {
            double t;
            if (strcmp(direction, "horizontal") == 0)
                t = (double)x / (double)w;
            else if (strcmp(direction, "diagonal") == 0)
                t = ((double)x / (double)w + (double)y / (double)h) / 2.0;
            else /* vertical */
                t = (double)y / (double)h;

            size_t i = ((size_t)y * w + x) * 3;
            img[i + 0] = (unsigned char)(c1[0] + (int)((c2[0] - c1[0]) * t));
            img[i + 1] = (unsigned char)(c1[1] + (int)((c2[1] - c1[1]) * t));
            img[i + 2] = (unsigned char)(c1[2] + (int)((c2[2] - c1[2]) * t));
        }
    }
    return img;
}

static unsigned char *make_pattern(int w, int h, int seed) {
    rng_state = (uint64_t)seed * 0x9E3779B97F4A7C15ULL;
    if (seed == 0) rng_state ^= (uint64_t)time(NULL);

    unsigned char *img = (unsigned char *)malloc((size_t)w * h * 3);
    if (!img) return NULL;

    /* Тёмная подложка */
    for (size_t i = 0; i < (size_t)w * h * 3; i++) img[i] = 12;

    int circles = brain_rand_range(20, 60);
    for (int c = 0; c < circles; c++) {
        int cx = brain_rand_range(-100, w + 100);
        int cy = brain_rand_range(-100, h + 100);
        int r = brain_rand_range(50, 300);
        int cr = brain_rand_range(0, 255);
        int cg = brain_rand_range(0, 255);
        int cb = brain_rand_range(0, 255);
        int alpha = brain_rand_range(30, 100);

        int x0 = cx - r; if (x0 < 0) x0 = 0;
        int x1 = cx + r; if (x1 > w) x1 = w;
        int y0 = cy - r; if (y0 < 0) y0 = 0;
        int y1 = cy + r; if (y1 > h) y1 = h;

        for (int y = y0; y < y1; y++) {
            for (int x = x0; x < x1; x++) {
                int dx = x - cx, dy = y - cy;
                if (dx * dx + dy * dy <= r * r) {
                    size_t i = ((size_t)y * w + x) * 3;
                    img[i + 0] = (unsigned char)((img[i + 0] * (255 - alpha) + cr * alpha) / 255);
                    img[i + 1] = (unsigned char)((img[i + 1] * (255 - alpha) + cg * alpha) / 255);
                    img[i + 2] = (unsigned char)((img[i + 2] * (255 - alpha) + cb * alpha) / 255);
                }
            }
        }
    }
    return img;
}

static int cmd_generate(int argc, char **argv) {
    const char *type = "gradient";
    const char *output = "wallpaper.png";
    int w = 1920, h = 1080;

    for (int i = 0; i < argc; i++) {
        if (strcmp(argv[i], "--type") == 0 && i + 1 < argc) type = argv[++i];
        else if (strcmp(argv[i], "--output") == 0 && i + 1 < argc) output = argv[++i];
        else if (strcmp(argv[i], "--width") == 0 && i + 1 < argc) w = atoi(argv[++i]);
        else if (strcmp(argv[i], "--height") == 0 && i + 1 < argc) h = atoi(argv[++i]);
    }

    unsigned char *img = NULL;

    if (strcmp(type, "gradient") == 0) {
        int c1[3] = {30, 15, 60}, c2[3] = {255, 100, 50};
        img = make_gradient(w, h, c1, c2, "vertical");
    } else if (strcmp(type, "dark") == 0) {
        int c1[3] = {10, 5, 30}, c2[3] = {40, 20, 80};
        img = make_gradient(w, h, c1, c2, "diagonal");
    } else if (strcmp(type, "pattern") == 0) {
        img = make_pattern(w, h, 0);
    } else {
        fprintf(stderr, "brain: неизвестный тип: %s\n", type);
        return 1;
    }

    if (!img) {
        fprintf(stderr, "brain: не хватило памяти\n");
        return 1;
    }

    if (!stbi_write_png(output, w, h, 3, img, w * 3)) {
        fprintf(stderr, "brain: не удалось записать %s\n", output);
        free(img);
        return 1;
    }
    free(img);

    StrBuf sb; sb_init(&sb);
    sb_append(&sb, "{\"status\":\"ok\",\"path\":");
    sb_append_json_string(&sb, output);
    sb_append(&sb, "}\n");
    printf("%s", sb.data);
    free(sb.data);
    return 0;
}

/* ============================================================
 *  Анализ файлов
 * ============================================================ */

typedef struct { const char *ext; const char *category; } SortRule;

static const SortRule SORT_RULES[] = {
    {"jpg", "Images"}, {"jpeg", "Images"}, {"png", "Images"},
    {"gif", "Images"}, {"bmp", "Images"}, {"webp", "Images"},
    {"svg", "Images"}, {"tiff", "Images"},
    {"pdf", "Documents/PDF"}, {"doc", "Documents/Word"}, {"docx", "Documents/Word"},
    {"txt", "Documents/Text"}, {"odt", "Documents/Text"}, {"rtf", "Documents/Text"},
    {"xls", "Documents/Spreadsheets"}, {"xlsx", "Documents/Spreadsheets"},
    {"csv", "Documents/Spreadsheets"},
    {"mp4", "Video"}, {"avi", "Video"}, {"mkv", "Video"},
    {"mov", "Video"}, {"wmv", "Video"}, {"webm", "Video"},
    {"mp3", "Audio"}, {"wav", "Audio"}, {"flac", "Audio"},
    {"ogg", "Audio"}, {"aac", "Audio"}, {"m4a", "Audio"},
    {"zip", "Archives"}, {"tar", "Archives"}, {"gz", "Archives"},
    {"rar", "Archives"}, {"7z", "Archives"},
    {"py", "Code/Python"}, {"js", "Code/JavaScript"}, {"ts", "Code/TypeScript"},
    {"rs", "Code/Rust"}, {"go", "Code/Go"}, {"lua", "Code/Lua"},
    {"c", "Code/C"}, {"cpp", "Code/C++"}, {"h", "Code/Headers"},
    {NULL, NULL}
};

static void str_to_lower(char *s) {
    for (; *s; s++) if (*s >= 'A' && *s <= 'Z') *s += 32;
}

/* Последний сегмент после '/' (или '\\' на Windows) */
static const char *path_base(const char *path) {
    const char *base = strrchr(path, '/');
#ifdef _WIN32
    const char *bslash = strrchr(path, '\\');
    if (bslash && (!base || bslash > base)) base = bslash;
#endif
    return base ? base + 1 : path;
}

/* Возвращает расширение файла (без точки, в нижнем регистре) */
static const char *file_extension(const char *name) {
    const char *dot = strrchr(name, '.');
    if (!dot) return "";
    return dot + 1;
}

static void categorize(const char *ext, const char **category, const char **folder) {
    char extl[32];
    snprintf(extl, sizeof(extl), "%s", ext);
    str_to_lower(extl);

    for (const SortRule *r = SORT_RULES; r->ext; r++) {
        if (strcmp(r->ext, extl) == 0) {
            *category = r->category;
            *folder = path_base(r->category);
            return;
        }
    }
    *category = "Other";
    *folder = "Other";
}

static void format_size(double bytes, char *out, size_t outlen) {
    const char *units[] = {"B", "KB", "MB", "GB", "TB"};
    int u = 0;
    while (bytes >= 1024.0 && u < 4) { bytes /= 1024.0; u++; }
    snprintf(out, outlen, "%.1f %s", bytes, units[u]);
}

/* Проверяет, является ли путь обычным файлом */
static int is_file(const char *path) {
#ifdef _WIN32
    DWORD attr = GetFileAttributesA(path);
    return attr != INVALID_FILE_ATTRIBUTES && !(attr & FILE_ATTRIBUTE_DIRECTORY);
#else
    struct stat st;
    if (stat(path, &st) != 0) return 0;
    return S_ISREG(st.st_mode);
#endif
}

static long long file_size(const char *path) {
#ifdef _WIN32
    /* Win32: GetFileAttributesExA возвращает размер без открытия дескриптора */
    WIN32_FILE_ATTRIBUTE_DATA fd;
    if (!GetFileAttributesExA(path, GetFileExInfoStandard, &fd)) return 0;
    return (long long)fd.nFileSizeLow | ((long long)fd.nFileSizeHigh << 32);
#else
    struct stat st;
    if (stat(path, &st) != 0) return 0;
    return (long long)st.st_size;
#endif
}

/* Добавляет JSON-объект анализа одного файла в буфер */
static void analyze_file_json(StrBuf *sb, const char *path) {
    char abs[4096];
#ifdef _WIN32
    if (_fullpath(abs, path, sizeof(abs)) == NULL) snprintf(abs, sizeof(abs), "%s", path);
#else
    if (path[0] == '/') {
        snprintf(abs, sizeof(abs), "%s", path);
    } else {
        /* Превращаем относительный путь в абсолютный через текущую директорию */
        if (getcwd(abs, sizeof(abs)) == NULL) snprintf(abs, sizeof(abs), "%s", path);
        size_t l = strlen(abs);
        if (l > 0 && l + strlen(path) + 1 < sizeof(abs)) {
            snprintf(abs + l, sizeof(abs) - l, "/%s", path);
        }
    }
#endif

    const char *name = path_base(abs);
    const char *ext = file_extension(name);
    const char *category, *folder;
    categorize(ext, &category, &folder);

    long long size = file_size(abs);
    char human[32];
    format_size((double)size, human, sizeof(human));

    sb_append(sb, "{\"name\":");       sb_append_json_string(sb, name);
    sb_append(sb, ",\"path\":");       sb_append_json_string(sb, abs);
    sb_append(sb, ",\"extension\":");  sb_append_json_string(sb, ext);
    sb_append(sb, ",\"size_bytes\":"); sb_printf(sb, "%lld", size);
    sb_append(sb, ",\"size_human\":"); sb_append_json_string(sb, human);
    sb_append(sb, ",\"category\":");   sb_append_json_string(sb, category);
    sb_append(sb, ",\"priority\":2");
    sb_append(sb, ",\"suggested_folder\":"); sb_append_json_string(sb, folder);
    sb_append(sb, "}");
}

/* Рекурсивно обходит директорию, заполняя JSON-массив. first=1 пока элемент первый. */
static void walk_directory(StrBuf *sb, const char *dirpath, int *first) {
#ifdef _WIN32
    char pattern[4096];
    snprintf(pattern, sizeof(pattern), "%s\\*", dirpath);
    WIN32_FIND_DATAA fd;
    HANDLE h = FindFirstFileA(pattern, &fd);
    if (h == INVALID_HANDLE_VALUE) return;
    do {
        if (strcmp(fd.cFileName, ".") == 0 || strcmp(fd.cFileName, "..") == 0) continue;
        char full[4096];
        snprintf(full, sizeof(full), "%s\\%s", dirpath, fd.cFileName);
        if (fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) {
            walk_directory(sb, full, first);
        } else {
            if (!*first) sb_append(sb, ",");
            *first = 0;
            analyze_file_json(sb, full);
        }
    } while (FindNextFileA(h, &fd));
    FindClose(h);
#else
    DIR *d = opendir(dirpath);
    if (!d) return;
    struct dirent *e;
    while ((e = readdir(d)) != NULL) {
        if (strcmp(e->d_name, ".") == 0 || strcmp(e->d_name, "..") == 0) continue;
        char full[4096];
        snprintf(full, sizeof(full), "%s/%s", dirpath, e->d_name);
        struct stat st;
        if (stat(full, &st) != 0) continue;
        if (S_ISDIR(st.st_mode)) {
            walk_directory(sb, full, first);
        } else if (S_ISREG(st.st_mode)) {
            if (!*first) sb_append(sb, ",");
            *first = 0;
            analyze_file_json(sb, full);
        }
    }
    closedir(d);
#endif
}

static int cmd_analyze(int argc, char **argv) {
    if (argc < 1) {
        fprintf(stderr, "brain: укажите путь для анализа\n");
        return 1;
    }
    const char *target = argv[0];
    StrBuf sb; sb_init(&sb);
    sb_append(&sb, "[");
    int first = 1;

    if (is_file(target)) {
        analyze_file_json(&sb, target);
    } else {
        walk_directory(&sb, target, &first);
    }

    sb_append(&sb, "]\n");
    printf("%s", sb.data);
    free(sb.data);
    return 0;
}

/* ============================================================
 *  Конвертация изображений
 * ============================================================ */

static int cmd_convert(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "brain: укажите input и output\n");
        return 1;
    }
    const char *input = argv[0];
    const char *output = argv[1];
    const char *fmt = "PNG";

    for (int i = 2; i < argc; i++) {
        if (strcmp(argv[i], "--format") == 0 && i + 1 < argc) fmt = argv[++i];
    }

    int w, h, n;
    unsigned char *data = stbi_load(input, &w, &h, &n, 3);
    if (!data) {
        fprintf(stderr, "brain: не удалось прочитать %s: %s\n", input, stbi_failure_reason());
        return 1;
    }

    int ok = 0;
    if (strcasecmp(fmt, "PNG") == 0)
        ok = stbi_write_png(output, w, h, 3, data, w * 3);
    else if (strcasecmp(fmt, "JPEG") == 0 || strcasecmp(fmt, "JPG") == 0)
        ok = stbi_write_jpg(output, w, h, 3, data, 95);
    else if (strcasecmp(fmt, "BMP") == 0)
        ok = stbi_write_bmp(output, w, h, 3, data);
    else if (strcasecmp(fmt, "TGA") == 0)
        ok = stbi_write_tga(output, w, h, 3, data);
    else
        fprintf(stderr, "brain: неизвестный формат: %s\n", fmt);

    stbi_image_free(data);

    if (!ok) {
        fprintf(stderr, "brain: не удалось записать %s\n", output);
        return 1;
    }

    StrBuf sb; sb_init(&sb);
    sb_append(&sb, "{\"status\":\"ok\",\"path\":");
    sb_append_json_string(&sb, output);
    sb_append(&sb, "}\n");
    printf("%s", sb.data);
    free(sb.data);
    return 0;
}

/* ============================================================
 *  Точка входа
 * ============================================================ */

static void usage(void) {
    fprintf(stderr,
        "Использование:\n"
        "  brain generate [--type gradient|pattern|dark] [--width W] [--height H] [--output FILE.png]\n"
        "  brain analyze <путь_к_файлу_или_папке>\n"
        "  brain convert <input> <output> [--format PNG|JPEG|BMP|TGA]\n");
}

int main(int argc, char **argv) {
    if (argc < 2) {
        usage();
        return 1;
    }

    const char *cmd = argv[1];

    if (strcmp(cmd, "generate") == 0)
        return cmd_generate(argc - 2, argv + 2);
    if (strcmp(cmd, "analyze") == 0)
        return cmd_analyze(argc - 2, argv + 2);
    if (strcmp(cmd, "convert") == 0)
        return cmd_convert(argc - 2, argv + 2);

    fprintf(stderr, "brain: неизвестная команда: %s\n", cmd);
    usage();
    return 1;
}